pyro/distributions/zero_inflated.py

```diff
@@ -0,0 +1,99 @@
+import torch
+from torch.distributions import constraints
+from torch.distributions.utils import broadcast_all, lazy_property
+
+from pyro.distributions import TorchDistribution, Poisson, NegativeBinomial
+
+
+class ZeroInflatedDistribution(TorchDistribution):
+    """
+    Base class for a Zero Inflated distribution.
+
+    :param torch.Tensor gate: probability of extra zeros given via a Bernoulli distribution.
+    :param TorchDistribution distribution: the secondary distribution.
+    """
+    arg_constraints = {"gate": constraints.unit_interval}
+
+    def __init__(self, gate, distribution, validate_args=None):
+        self.gate = gate
+        self.distribution = distribution
+        batch_shape = self.gate.shape
+        event_shape = torch.Size()
+
+        super(ZeroInflatedDistribution, self).__init__(
+            batch_shape, event_shape, validate_args
+        )
+
+    def log_prob(self, value):
+        if self._validate_args:
+            self._validate_sample(value)
+
+        gate, value = broadcast_all(self.gate, value)
+        log_prob = (-gate).log1p() + self.distribution.log_prob(value)
+        log_prob = torch.where(value == 0, (gate + log_prob.exp()).log(), log_prob)
+        return log_prob
+
+    def sample(self, sample_shape=torch.Size()):
+        shape = self._extended_shape(sample_shape)
+        with torch.no_grad():
+            mask = torch.bernoulli(self.gate.expand(shape)).bool()
+            samples = self.distribution.expand(shape).sample()
+            samples = torch.where(mask, samples.new_zeros(()), samples)
+        return samples
+
+    @lazy_property
+    def mean(self):
+        return (1 - self.gate) * self.distribution.mean
+
+    @lazy_property
+    def variance(self):
+        return (1 - self.gate) * (
+            self.distribution.mean ** 2 + self.distribution.variance
+        ) - (self.mean) ** 2
+
+    def expand(self, batch_shape):
```

Comment:
@kashif I think you would need to follow PyTorch convention here:
```
    def expand(self, batch_shape, _instance=None):
        new = self._get_checked_instance(type(self), _instance)
        batch_shape = torch.Size(batch_shape)
        gate = self.gate.expand(batch_shape)
        distribution = self.distribution.expand(batch_shape)
        super(type(self), new).__init__(gate, distribution, validate_args=False)
        new._validate_args = self._validate_args
        return new
```

I would recommend to change `distribution` to `base_dist`, which is also a convention name in Pyro. :)

Reply:
thanks will do!